syz-manager/snapshot.go

```diff
@@ -0,0 +1,200 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"time"
+	"unsafe"
+
+	"github.com/google/flatbuffers/go"
+	"github.com/google/syzkaller/pkg/flatrpc"
+	"github.com/google/syzkaller/pkg/fuzzer/queue"
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/vm"
+)
+
+func (mgr *Manager) snapshotLoop() {
+	queue.StatNumFuzzing.Add(mgr.vmPool.Count())
+
+	reqc := make(chan *queue.Request, 2*mgr.vmPool.Count())
+	go func() {
+		for {
+			reqc <- mgr.source.Next()
+		}
+	}()
+
+	errc := make(chan error)
+	for index := 0; index < mgr.vmPool.Count(); index++ {
+		index := index
+		go func() {
+			errc <- mgr.snapshotVM(index, reqc)
+		}()
+	}
+	log.Fatal(<-errc)
+}
+
+func (mgr *Manager) snapshotVM(index int, reqc <-chan *queue.Request) error {
+	inst, err := mgr.vmPool.Create(index)
+	if err != nil {
+		return err
+	}
+	executor, err := inst.Copy(mgr.cfg.ExecutorBin)
+	if err != nil {
+		return err
+	}
+	// All network connections (including ssh) will break once we start restoring snapshots.
+	// So we start a background process and log to /dev/kmsg.
+	cmd := fmt.Sprintf("nohup %v exec snapshot 1>/dev/null 2>/dev/kmsg </dev/null &", executor)
+	if _, _, err := inst.Run(time.Hour, mgr.reporter, cmd); err != nil {
+		return err
+	}
+
+	for first := true; ; first = false {
+		var req *queue.Request
+		select {
+		case req = <-reqc:
+		default:
+			start := time.Now()
+			req = <-reqc
+			queue.StatNoExecRequests.Add(1)
+			queue.StatNoExecDuration.Add(int(time.Since(start)))
+		}
+
+		if first {
+			if err := mgr.snapshotSetup(inst, req.ExecOpts.EnvFlags); err != nil {
```

Comment:
Nit: should we also add some panic in case of an unexpected change of `EnvFlags` between different requests?

Reply:
Added a check.